### PR TITLE
[NFC] Replace miopen::ProblemDescription with conv::ProblemDescription, part 3

### PR DESCRIPTION
**Pull request by averinevg**
* Make miopen::ProblemDescription derived from conv::ProblemDescription
* miopen::ProblemDescription: remove getters that exist in the parent class
* conv::ProblemDescription: use proper data type of return value for getters that use data from TensorDescriptor
* conv::ProblemDescription: add trailing underscores to getters with changed data type in order to get the attention of developers who are not familiar with these changes (will be removed after ROCm 6.0)
* Remove unused code

### CODE REVIEW

**Comment by averinevg** [comment]
@atamazov Please review

**Comment by atamazov** [comment]
@averinevg 
> conv::ProblemDescription: add trailing underscores to getters with changed data type in order to get the attention of developers who are not familiar with these changes (will be removed after ROCm 6.0)

Are you going to remove those underscores after MIOpen 6.0 RC is cut off?

File: src/include/miopen/tensor.hpp
Comment (atamazov) [inline src/include/miopen/tensor.hpp:288]: [Q] why this is necessary?

File: src/include/miopen/conv/problem_description.hpp
Comment (atamazov) [inline src/include/miopen/conv/problem_description.hpp:203]: [Q] I am not sure this is a good change. Let's discuss.

**Review by atamazov — COMMENTED** [review]
Something to discuss.

**Comment by averinevg** [comment]
> @averinevg
> 
> > conv::ProblemDescription: add trailing underscores to getters with changed data type in order to get the attention of developers who are not familiar with these changes (will be removed after ROCm 6.0)
> 
> Are you going to remove those underscores after MIOpen 6.0 RC is cut off?

Yes, this is a temporary change

File: src/include/miopen/tensor.hpp
Comment (averinevg) [reply src/include/miopen/tensor.hpp:288]: This variable is always positive and using a signed data type can be misleading

Comment (atamazov) [reply src/include/miopen/tensor.hpp:288]: That is, you just caught a not quite high-quality code and decided to fix it, right? 😄

**Comment by atamazov** [comment]
WRT https://github.com/ROCmSoftwarePlatform/MIOpen/pull/2303#discussion_r1286316295. After ~2 hours discussion we decided to continue as is and then switch to size_t in a dedicated PR.

File: src/solver/conv_MP_bidirectional_winograd.cpp
Comment (DrizztDoUrden) [inline src/solver/conv_MP_bidirectional_winograd.cpp:92]: An idea for future work to ease solver developers life:
Similar code is verbose and present in most of our solvers. We should add some method like `CommonlyUsedConvParameters GetCommonlyUsedConvParameters` with `CommonlyUsedConvParameters` being some type which is compatible to structured bindings so it would be possible to either extract them as `const auto [grop_cnt, N, K, ...] = problem.GetCommonlyUsedConvParameters()` or just as `const auto params = GetCommonlyUsedConvParameters()` to use as `params.group_cnt` and such.

**Review by DrizztDoUrden — DISMISSED** [review]
LGTM

File: src/include/miopen/tensor.hpp
Comment (averinevg) [reply src/include/miopen/tensor.hpp:288]: Yes, you are right :smile:. I decided to change this too, because it is related to my changes

File: src/solver/conv_MP_bidirectional_winograd.cpp
Comment (averinevg) [reply src/solver/conv_MP_bidirectional_winograd.cpp:92]: Perhaps we should also refactor other "helpers" that get data from `ProblemDescription`. I assume that they partially overlap and can be generalized.

**Review by junliume — DISMISSED** [review]
LGTM but I am waiting for the good opportunity to merge. Too many files changed means higher chances for other PRs to have conflicts.

**Review by DrizztDoUrden — DISMISSED** [review]

**Review by JehandadKhan — APPROVED** [review]

**Comment by JehandadKhan** [comment]
@junliume  We can merge this when you get a chance

**Review by junliume — APPROVED** [review]

File: src/include/miopen/conv/problem_description.hpp
Comment (amberhassaan) [reply src/include/miopen/conv/problem_description.hpp:203]: I would also like to know why we went to `unsigned` from `size_t` and why the names have an `_` at the end of getters.

Comment (averinevg) [reply src/include/miopen/conv/problem_description.hpp:203]: @amberhassaan You can find details here https://github.com/ROCmSoftwarePlatform/MIOpen/pull/2303#issue-1839780618 and here https://github.com/ROCmSoftwarePlatform/MIOpen/pull/2303#issuecomment-1672048340

In most places in the library `miopen::ProblemDescription` is used, its getters returned the `int` type, the getters of `conv::ProblemDescription` returned the `size_t` type. For now, it is enough for us to have `int`. `unsigned` is used because we don't need a sign.

File: src/include/miopen/problem_description.hpp
Comment (amberhassaan) [inline src/include/miopen/problem_description.hpp:41]: Is this flag documented somewhere?

Comment (averinevg) [reply src/include/miopen/problem_description.hpp:41]: No, it is supposed to be used with MIFin for backward compatibility and will be removed soon.

**Pull request merged**
